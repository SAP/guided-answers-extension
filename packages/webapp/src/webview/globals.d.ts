interface Window {
    vscode: {
        postMessage: (message: string | JSON) => void;
    };
}
