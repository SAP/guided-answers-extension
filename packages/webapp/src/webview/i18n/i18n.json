{
    "GUIDED_ANSWER_HEADER": "Guide",
    "SEARCH_GUIDED_ANSWERS": "Search Guided Answers",
    "ALL_ANSWERS": "Home",
    "STEP_BACK": "Step back",
    "RESTART": "Restart",
    "GUIDED_ANSWERS": "Guided Answers",
    "NO_ANSWERS_FOUND": "No answers found",
    "PLEASE_MODIFY_SEARCH": "Please modfiy search",
    "THIS_SOLVED_MY_ISSUE": "This solved my issue",
    "THIS_DID_NOT_SOLVED_MY_ISSUE": "This did not solve my issue",
    "THANK_YOU_TEXT": "We are glad to hear that your issue has been resolved and we hope you enjoy using the Guided Answers application.",
    "THANKS": "Thanks!",
    "HOME": "Home",
    "SEND": "Send",
    "CLOSE": "Close",
    "ISSUE_IS_NOT_RESOLVED": "Issue is not resolved",
    "WE_ARE_SORRY_TO_HEAR_THAT_YOUR_ISSUE_IS_NOT_YET_RESOLVED": "We are sorry to hear that your issue is not yet resolved.",
    "PLEASE_TELL_US_IF_THIS_ANSWER_WAS_HELPFUL": "Please tell us if this answer was helpful",
    "THERE_ARE_SEVERAL_OPTIONS_FOR_GETTING_FURTHER_ASSISTANCE": "There are several options for getting further assistance:",
    "SEARCH_FOR_ANOTHER_GUIDED_ANSWER": "Search for another Guided Answer",
    "START_AN_EXPERT_CHAT": "Start an Expert Chat",
    "SCHUEDULE_AN_EXPERT": "Schedule an Expert",
    "OPEN_AN_INCIDENT": "Open an Incident",
    "ASK_THE_SAP_COMMUNITY": "Ask the SAP Community",
    "FEEDBACK": "Feedback",
    "FEEDBACK_DIALOG_TITLE": "Is this content helpful?",
    "FEEDBACK_DIALOG_SUBTEXT": "If you have suggestions on how to improve this content we would love to hear them!",
    "FEEDBACK_DIALOG_DISCLAIMER": "Your feedback is anonymous, we do not collect any personal data.",
    "FEEDBACK_DIALOG_SUGGESTION": "Your suggestion",
    "FEEDBACK_SENT_DIALOG_TITLE": "Message sent",
    "FEEDBACK_SENT_DIALOG_SUBTEXT": "Thanks for the feedback!"
}
