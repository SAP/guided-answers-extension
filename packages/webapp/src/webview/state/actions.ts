export {
    updateGuidedAnswerTrees,
    updateActiveNode,
    selectNode,
    goToPreviousPage,
    goToAllAnswers,
    restartAnswer,
    executeCommand,
    setActiveTree,
    searchTree,
    setQueryValue,
    webviewReady,
    getBetaFeatures,
    sendFeedbackOutcome,
    sendFeedbackComment,
    guideFeedback,
    setProductFilters,
    setComponentFilters,
    resetFilters,
    setPageSize,
    feedbackStatus,
    feedbackResponse
} from '@sap/guided-answers-extension-types';
