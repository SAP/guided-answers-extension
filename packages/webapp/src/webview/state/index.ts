export { actions, store } from './store';
export { getInitialState } from './reducers';
