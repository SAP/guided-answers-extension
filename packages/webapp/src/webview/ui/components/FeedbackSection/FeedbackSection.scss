.feedback-container {
    h3 {
        color: var(--vscode-settings-headerForeground);
        font-size: 14px;
        font-weight: 700;
    }

    margin-bottom: 15px;
}

.feedback-subcontainer {
    display: flex;
    justify-content: flex-start;
    flex-direction: row;
    gap: 20px;
    width: 100%;
}

button:focus-visible {
    outline: 1px solid var(--vscode-focusBorder, #007fd4);
}

.feedback-box {
    border: 2px solid var(--vscode-terminal-ansiGreen);
    border-radius: 3px;
    display: flex;
    flex-direction: row;
    padding: 20px;
    align-items: center;
    gap: 12px;
    overflow-wrap: break-word;
    width: 50%;
    text-align: left;
}

.solved-hover:hover {
    background-color: var(--vscode-terminal-ansiGreen);
    cursor: pointer;
    > h3 {
        color: var(--vscode-editor-background);
    }
    svg > path,
    svg > rect {
        fill: var(--vscode-editor-background);
    }
}

.not-solved-hover:hover {
    background-color: var(--vscode-terminal-ansiRed);
    cursor: pointer;
    > h3 {
        color: var(--vscode-editor-background);
    }
    svg > path,
    svg > rect {
        fill: var(--vscode-editor-background);
    }
}

@media only screen and (max-width: 625px) {
    .feedback-box > svg {
        display: none;
    }
}
