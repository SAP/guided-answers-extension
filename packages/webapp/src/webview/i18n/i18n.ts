import { use } from 'i18next';
import { initReactI18next } from 'react-i18next';
import * as i18nEn from './i18n.json';

export { i18n } from 'i18next';

export function initI18n(): void {
    use(initReactI18next)
        .init({
            resources: {
                en: {
                    translation: i18nEn
                }
            },
            lng: 'en',
            fallbackLng: 'en',
            interpolation: {
                escapeValue: false
            }
        })
        .catch((error) => console.error(error));
}
