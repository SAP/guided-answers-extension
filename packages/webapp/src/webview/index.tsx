import React from 'react';
import { render } from 'react-dom';
import { Provider } from 'react-redux';
import { App } from './ui/components/App';
import { actions, store } from './state';
import { initI18n } from './i18n';

/**
 * Initialization of i18n, icons
 */
initI18n();

/**
 * Render app
 */
render(
    <Provider store={store}>
        <App />
    </Provider>,
    document.getElementById('root') as HTMLElement
);

/**
 * Fire action which is captured in hosting app (ide-extension) to signal that we are ready to receive messages
 */
actions.webviewReady();

/**Once webview is ready we focus to search field */
document.getElementById('search-field')?.focus();
