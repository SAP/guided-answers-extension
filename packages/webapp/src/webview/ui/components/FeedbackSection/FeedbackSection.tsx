import React, { ReactElement } from 'react';
import { useSelector } from 'react-redux';
import NotSolved from './images/not-solved.svg';
import Solved from './images/solved.svg';
import i18next from 'i18next';
import { actions } from '../../../state';
import './FeedbackSection.scss';
import { MessageDialogBox } from '../DialogBoxes/MessageDialogBox';
import type { AppState } from '../../../types';
import type { GuidedAnswerNodeId, GuidedAnswerTreeId } from '@sap/guided-answers-extension-types';
import { FocusZone, FocusZoneDirection } from '@fluentui/react-focus';

/**
 * @returns - react element with the feedback section
 */
export function FeedbackSection(): ReactElement {
    const treeId = useSelector<AppState, GuidedAnswerTreeId>((state) => state.activeGuidedAnswer!.TREE_ID);
    const nodeId = useSelector<AppState, GuidedAnswerNodeId>((state) => state.activeGuidedAnswerNode[0].NODE_ID);
    let guideFeedback = useSelector<AppState, boolean | null>((state) => state.guideFeedback);
    return (
        <div className="feedback-container">
            <h3>{i18next.t('PLEASE_TELL_US_IF_THIS_ANSWER_WAS_HELPFUL')}</h3>
            <FocusZone direction={FocusZoneDirection.horizontal} className="feedback-subcontainer" role="tree">
                <button
                    className="feedback-box solved-hover"
                    onClick={(): void => {
                        actions.guideFeedback(true);
                        actions.sendFeedbackOutcome({ treeId, nodeId, solved: true });
                    }}>
                    <Solved />
                    <h3>{i18next.t('THIS_SOLVED_MY_ISSUE')}</h3>
                </button>
                <button
                    className="feedback-box not-solved-hover"
                    style={{ border: '2px solid var(--vscode-terminal-ansiRed)' }}
                    onClick={(): void => {
                        actions.guideFeedback(false);
                        actions.sendFeedbackOutcome({ treeId, nodeId, solved: false });
                    }}>
                    <NotSolved />
                    <h3>{i18next.t('THIS_DID_NOT_SOLVED_MY_ISSUE')}</h3>
                </button>
            </FocusZone>
            <MessageDialogBox
                dialogTitle={i18next.t('THANKS')}
                dialogText={i18next.t('THANK_YOU_TEXT')}
                dialogVisible={guideFeedback === true ? true : false}
                primaryButtonAction={() => actions.goToAllAnswers()}
                defaultButtonAction={() => actions.guideFeedback(null)}
                stylingClassName="solved-message-dialog"
            />
        </div>
    );
}
