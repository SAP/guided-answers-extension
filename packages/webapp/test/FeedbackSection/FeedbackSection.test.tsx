import React from 'react';
import { shallow } from 'enzyme';
import { actions } from '../../src/webview/state';
import { FeedbackSection } from '../../src/webview/ui/components/FeedbackSection/FeedbackSection';
import { initI18n } from '../../src/webview/i18n';

jest.mock('../../src/webview/state', () => {
    return {
        actions: {
            guideFeedback: jest.fn(),
            sendFeedbackOutcome: jest.fn()
        }
    };
});

jest.mock('react-redux', () => ({
    ...jest.requireActual('react-redux'),
    useSelector: jest.fn().mockReturnValue({ guideFeedback: null })
}));

describe('Feedback Section component', () => {
    let wrapper: any;
    initI18n();
    beforeEach(() => {
        wrapper = shallow(<FeedbackSection />);
    });

    afterEach(() => {
        jest.clearAllMocks();
        wrapper.unmount();
    });

    it('render component Feedback Section', () => {
        let component = wrapper.html();
        expect(component).toMatchInlineSnapshot(
            `"<div class="feedback-container"><h3>Please tell us if this answer was helpful</h3><div class="ms-FocusZone css-108 feedback-subcontainer" role="tree" data-focuszone-id="FocusZone0"><button class="feedback-box solved-hover"><svg></svg><h3>This solved my issue</h3></button><button class="feedback-box not-solved-hover" style="border:2px solid var(--vscode-terminal-ansiRed)"><svg></svg><h3>This did not solve my issue</h3></button></div></div>"`
        );
    });

    it('clicking on solved message should change state', () => {
        wrapper.find('.feedback-box').at(0).simulate('click');
        expect(actions.guideFeedback).toBeCalled();
    });

    it('clicking on not solved message should change state', () => {
        wrapper.find('.feedback-box').at(1).simulate('click');
        expect(actions.guideFeedback).toBeCalled();
    });
});
