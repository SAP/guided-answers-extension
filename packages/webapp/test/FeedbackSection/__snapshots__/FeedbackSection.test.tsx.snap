// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feedback Section component Should render a FeedbackSection component 1`] = `
<div>
  <div
    class="feedback-container"
  >
    <h3>
      Please tell us if this answer was helpful
    </h3>
    <div
      class="ms-FocusZone css-110 feedback-subcontainer"
      data-focuszone-id="FocusZone0"
      role="tree"
    >
      <button
        class="feedback-box solved-hover"
        id="solved-issue-button"
        tabindex="0"
      >
        <svg
          focusable="false"
        />
        <h3>
          This solved my issue
        </h3>
      </button>
      <button
        class="feedback-box not-solved-hover"
        id="not-solved-issue-button"
        tabindex="-1"
      >
        <svg
          focusable="false"
        />
        <h3>
          This did not solve my issue
        </h3>
      </button>
    </div>
  </div>
</div>
`;
